### ISSUE TITLE
store.dispatch(...).subscribe(...) called twice for async Actions

### ISSUE THREAD

**Issue by xmlking** [amcdnl/ngxs#104]
store.dispatch(...).subscribe(...) called twice when Action with `async` method implementation dispatched. 
This is leading bugs as we have some business logic executed twice with in `subscribe` method.

if you remove `async` , this will print once as expected. 

This issue got introduced in 2.0 release. it was working fine in 1.x 

**stackblitz:**
https://stackblitz.com/edit/ngxs-router-sumo

```ts
import { State, Action, StateContext } from 'ngxs';

export class Add {}

@State({
  name:'app',
  defaults: {
    count: 0
  }
})
export class AppState {
  @Action(Add)
  async add({ getState, setState }: StateContext<any>) {
    await new Promise(resolve => setTimeout(resolve, 1000));
    console.log("in add")
    setState({
      count: getState().count + 1
    })
  }
}
```

```ts
  onClick() {
    this.store.dispatch(new Add()).subscribe( () => {
      console.log('... this will print twice')
    });
  }
```

**Comment by deebloo** [comment]
@xmlking can you confirm this only happens twice this way? it doesn't happen when not using way or when using an observable?

**Comment by xmlking** [comment]
It  happened when “async” is used. If I remove async it worked normally.  I haven’t tested observable  yet. Will do ...

**Comment by yarrgh** [comment]
It happens with an observable as well. Even in the dev tools I see double entries (sometimes triple) entries for my async actions. I'm also seeing "Object" entries. No idea what they are why they are there. On top of this, the entries in the dev tools are out of order. The action being dispatched comes after a dispatched action inside another. 

Example: I subscribe a dispatched `CreateRecipe` action inside my component (only dispatched once). The action then sets initial state (pending: true) and then calls the backend via observable form the `RecipeService`. It then dispatches a `CreateRecipeSuccess` with the created object to set the state otherwise it dispatches `CreateRecipeFail`. In the dev tools, `CreateRecipeSuccess` shows up first, followed by two entries of "Object" and then 3 entries of `CreateRecipe`. 

![image](https://user-images.githubusercontent.com/1577822/37664929-c895caae-2c21-11e8-89b7-cec184caba51.png)

**Comment by deebloo** [comment]
ok I can reproduce. @amcdnl it is whenever the action RETURNS a promise or an observable. calling setState inside inside of an async thing does not cause it. I know where to look now

**Comment by amcdnl** [comment]
Ya I think i know....

**Comment by amcdnl** [comment]
@jschwarty ^ 

> I think it’s because the observable you are creating in store.dispatch is unicast. So each subscribe gets its own execution context. Maybe that needs to be a multicast (like a Subject or BehaviorSubject)

**Comment by amcdnl** [comment]
I believe this fixes it - https://github.com/amcdnl/ngxs/commit/ce4e6638fb2d88dffdf49c5bc063c72109f097bf

@jschwarty - can u sanity check this for me?

**Comment by yarrgh** [comment]
That mostly fixed it. However, this doesn't fix subscribing to the observable returned from Store.dispatch(). If you subscribe to that observable, it causes the action to be dispatched again.

**Comment by yarrgh** [comment]
I guess this may be a different issue. I'll open up a new issue.

**Comment by uberpwner** [comment]
Hi @yarrgh 
Having same issue. If I'm subscribing to observable from Store.dispatch() I see my action twice in a log:
![2018-05-22_01-09-57](https://user-images.githubusercontent.com/15633416/40332484-ebe42336-5d5c-11e8-8d73-9bdc4278741d.png)

However `console.log();` inside action prints in console only once.

**Comment by conradbekondo** [comment]
This is still an issue in version 18.0.0

![image](https://github.com/ngxs/store/assets/112887424/8d987974-4851-4af5-9b34-2177ca6cba8c)

**Comment by dpietruc** [comment]
Hi,
Please make sure that the action types in your actions.ts are all unique.
In my case, I had double actions dispatch because two actions had the same type assigned.
